src/Enzo/enzo_EnzoReconstructorPLM.hpp

```diff
@@ -0,0 +1,406 @@
+// See LICENSE_CELLO file for license and copyright information
+
+/// @file     enzo_EnzoReconstructorPLM.hpp
+/// @author   Matthew Abruzzo (matthewabruzzo@gmail.com)
+/// @date     Wed May 1 2019
+/// @brief    [\ref Enzo] Implementation of Enzo's Piecewise Linear
+///           Reconstruction
+
+#ifndef ENZO_ENZO_RECONSTRUCTOR_PLM_HPP
+#define ENZO_ENZO_RECONSTRUCTOR_PLM_HPP
+
+#include <type_traits>
+
+/// @typedef limiter_function_call_signature
+/// @brief   This is the function call signature of the function call operator,
+///          `operator()`, that is expected for a `Limiter` functor that is
+///          passed to `EnzoReconstructorPLM`.
+///
+/// We expect the `operator()` method of a `Limiter` functor to be declared as:
+///
+/// @code
+///     enzo_float operator() (enzo_float vm1, enzo_float v, enzo_float vp1,
+///                            enzo_float theta_limiter);
+/// @endcode
+///
+/// The function should compute the limited slope to use in the piecewise
+/// linear reconstruction of the left and right interface values (which are
+/// ultimately used by the Riemann Solver) for some cell-centered variable `w`.
+/// The function is passed 3 contiguous values of `w`: `w_{i-1}`, `w_i`, and
+/// `w_{i+1} as the arguments `vm1`, `v` and `vp1`. The `theta_limiter`
+/// argument can be used to tune the limiter (or it can be ignored). The
+/// resulting limited slope, `dw` is used to reconstruct the interface
+/// variables in the equations:
+///     `w_{L,i+1/2} = w_i + 0.5*dw` and `w_{R,i+1/2} = w_i - 0.5*dw`
+
+typedef enzo_float (*limiter_function_call_signature)
+  (enzo_float, enzo_float, enzo_float, enzo_float);
+
+//----------------------------------------------------------------------
+
+/// @def      DEFINE_HAS_INSTANCE_METHOD
+/// @brief    Macro for defining a struct that is used to check if a given
+///           class definition has a public instance method with a specified
+///           name and function signature
+///
+/// @param traitsName The name of the struct that will be defined
+/// @param func_name  Name of the public instance method to check for
+/// @param signature  The signature that the method should have (this can
+///     be a function pointer).
+///
+/// This macro is modified slightly modified from the answer provided at
+/// https://stackoverflow.com/a/16824239 . The particular answer that this is
+/// adapted from allows the required instance method to be inherited. It is
+/// nominally used to define structs that can be used to explicitly check
+/// expectations about template arguments so that useful error messages can
+/// be provided for debugging.
```

Comment:
This macro allows me to define `has_limiter_function_call_operator` on line 96. The usage of the struct is very similar to something like `std::is_default_constructible`. I actually make use of this struct to write a `static_assert` on line 137.

For the sake of example, I'll try to compile the code after deleting the the `theta_limiter` argument from the `PLM_AthenaLimiter::operator()` method.

Error Messages using gcc:
1. When I don't have the `static_assert`: [output](https://pastebin.com/EBtGU0X7)
2. When I do have the `static_assert`: [output](https://pastebin.com/X34gWY29)

Error Messages using icc:
1. When I don't have the `static_assert`: [output](https://pastebin.com/D4FzP5QL)
2. When I do have the `static_assert`: [output](https://pastebin.com/hdMXgvKu)

In summary, you're absolutely right. This is overkill. I'll definitely remove it from here.

However, I think I also use this macro for defining similar checks with the Riemann Solvers. It's been a while, so I'm a little hazy on the details, but I think I was originally inspired to make these checks explicitly in that case because there were a couple of template usage. I'm going to go through a similar exercise again, to see if this is actually helpful in that case.

Reply:
Now to repeat the exercise for `EnzoRiemannImpl` (the `static_assert` statement is on line 150 of `enzo_EnzoRiemannImpl.hpp`). In this situation, I'm going to drop the `gamma` argument from the `HLLImpl::operator()` method.

Error Messages using gcc:
1. When I don't have the `static_assert`: [output](https://pastebin.com/xaUht80Z)
2. When I do have the `static_assert`: [output](https://pastebin.com/1D0QsnCN)

Error Messages using icc:
1. When I don't have the `static_assert`: [output](https://pastebin.com/LU8D9E0i)
2. When I do have the `static_assert`: [output](https://pastebin.com/sskD5ynV)

From a quick glance, I don't think that `static_assert` is making anything more obvious. I'll come back to this in ~hour (I got to go take care of something).

Reply:
I took another look at things and am kind of torn about this extra check in the case of the Riemann Solver.
- On the one hand, if you are modifying these functions, it may be obvious to you from the error message without the `static_assert` what's going wrong. 
- On the other hand, I think the `static_assert` makes the problem more obvious.

I am somewhat inclined to keep the `static_assert` for the Riemann Solver, but I'm happy to defer to your judgement, @forrestglines

To reiterate, I'm definitely going to remove the extra check from the Reconstructor. If we do keep the check for the Riemann Solver, then I'll probably move the macro to `enzo_EnzoRiemannUtils.hpp` (or maybe `enzo_EnzoRiemannImpl.hpp`)

Reply:
I think the `static_assert` is fine to leave in.